### DISCUSSION TITLE
Port SpecFlow.Actions to Reqnroll

### DISCUSSION THREAD

**Discussion by gasparnagy** [reqnroll/Reqnroll#27] [Ideas]
Originally I thought to delay porting it and combine it with a deeper review & revise, but it seems that many people use it. E.g. SpecFlow.Actions.Selenium was downloaded 11k times in the last 6 weeks.

So maybe we sohuld just port this as is, to make migration easier.

**Comment by tula-engineered-quality-hub** [comment]
Specflow actions are very useful in e2e tests!Specflow actions are very useful in e2e tests! facilitates execution logic in different browsers, urls, among others!

@gasparnagy , i have been working with e2e automation for a few years, on large projects. I ended up creating a private Library that abstracts all Selenium logic, with useful commands including, in addition to considering Specflow and Specflow Actions in these engines, if it is of interest to Reqnroll I can adjust this Lib for Reqnroll and make it available here, committing to continue maintaining this lib.

**Comment by SabotageAndi** [comment]
FYI: The SpecFlow.Actions.Selenium is still using Selenium 3. This would be a good chance to upgrade it to Selenium 4.

**Reply by tula-engineered-quality-hub** [reply]
Hello @SabotageAndi, really a great opportunity, this lib I created I already updated to 4 about 18 months ago, it was really worth it.

**Comment by gasparnagy** [comment]
OK. I started porting SpecFlow.Actions to https://github.com/reqnroll/Reqnroll.Actions, but the situation is not that good.

It seems that SpecFlow.Actions is super-legacy and it was not even ported yet to SpecFlow v4 and it does some magic that is not compatible with SpecFlow v4 as it is.

This is a huge code base and I don't think we will have the capacity to maintain all of it (at least not from the beginning). 

My suggestion is the following:
- I try to make it somehow compilable and working with Reqnroll.
- We release it as `Reqnroll.SpecFlowCompatibility.Actions.XXX`
- In parallel to that, we can think about re-building the most important elements of it as `Reqnroll.Actions.XXX`, e.g. the one with selenium. We can consider the get input from the lib of @tulavalle.

@tulavalle is your lib (with SpecFlow) on GitHub somewhere so that we can look?

**Reply by SabotageAndi** [reply]
What breaking changes did we introduce in v4 that the action plugins aren't easy to upgrade?

**Reply by gasparnagy** [reply]
Async

**Reply by gasparnagy** [reply]
The Configuration action contained a full copy of the test class generator.

**Reply by tula-engineered-quality-hub** [reply]
Hi @gasparnagy,  i temporarily put the project as public on my github: https://github.com/tulavalle/SpecflowSelenium.git . It is a project implemented with a QA perspective with a focus on productivity and reducing the learning curve in e2e automation, abstracting several repetitive use commands from Selenium, but considering important points of Specflow, in terms of opening the browser and Browser according to Specflow Actions, based on parameters in the test design. It is compatible with Selenium grid or without the grid, this is also parameterized in the test project. Regarding LivingDoc, screenshot generation logic is implemented in a folder in the project itself or in the Amazon bucket, the implementation is to make the image available in LivingDoc. There is also logic to automatically fail scenarios with bug or development tags, aiming not to execute a scenario with a reported bug and thinking about TDD, where tests start failing, in addition to allowing you to manage manual scenarios in relation to manual tests passing or failing based on the tag bug displaying the bug report link in LivingDoc.

If the project is of interest to Reqnroll, adjustments are likely to be necessary, including translation of the documentation. or seek the best of both worlds in relation to this project and the SpecFlow.Actions.Selenium.


Regarding the super legacy of Specflow.Actions, before it I used the concept of targets that I had in SpecRun, for e2e tests it worked very well, but technically I don't know how much it would cost in relation to adjustments to the Actions project.

**Reply by gasparnagy** [reply]
@tulavalle thx. I will review. It seems I was able to get the targets concept working, so that will be available for sure.

**Reply by SabotageAndi** [reply]
> The Configuration action contained a full copy of the test class generator.

Yeah, now I remember it again. That was needed because the needed API change in SpecFlow would trigger a new Minor Version release, what we didn't wanted to do at the time.

**Reply by gasparnagy** [reply]
Fortunately I have found an easy workaround that avoids this. https://github.com/reqnroll/Reqnroll.Actions/commit/4cc76fa4ff422c50e481ef5351236a68f7d57fe4

**Comment by gasparnagy** [comment]
The current state of the SpecFlow "Actions" packages has been also ported to make the migration easier as `Reqnroll.SpecFlowCompatibility.Actions.*` (e.g. [`Reqnroll.SpecFlowCompatibility.Actions.Selenium`](https://www.nuget.org/packages/Reqnroll.SpecFlowCompatibility.Actions.Selenium)). Once they are rewised and updated they will be moved to "Reqnroll.Actions.*".

The migration guide has been updated accordingly.

To track the concept revise, I created a separate discussion https://github.com/orgs/reqnroll/discussions/34. @tulavalle we will continue the discussion about the selenium stuff there.

I close this topic now.
